apps/cbportal/base/cloudbroker__machine/methodclass/cloudbroker_machine.py

```diff
@@ -103,6 +104,35 @@ def start(self, machineId, reason, **kwargs):
             j.apps.cloudbroker.machine.untag(vmachine.id, "start")
         self.cb.actors.cloudapi.machines.start(machineId=machineId)
 
+    @auth(['level1', 'level2', 'level3'])
+    @wrap_remote
+    def restore(self, machineId, reason, **kwargs):
+        """
+        Restore a deleted machine
+
+        :param machineId: id of the machine
+        """
+        ctx = kwargs['ctx']
+        ctx.env['JS_AUDIT'] = True
+        ctx.env['tags'] += " machineId:{}".format(machineId)
+        machine = self.models.vmachine.get(int(machineId))
+        cloudspace = self.models.cloudspace.get(machine.cloudspaceId)
+        if machine.status != resourcestatus.Machine.DELETED:
+            raise exceptions.BadRequest("Can't restore a non deleted machine")
+        with self.models.cloudspace.lock('{}_ip'.format(machine.cloudspaceId)):
+            nic = machine.nics[0]
+            nic.type = 'bridge'
+            nic.ipAddress = self.cb.cloudspace.network.getFreeIPAddress(cloudspace)
+            machine.status = resourcestatus.Machine.HALTED
+            machine.updateTime = int(time.time())
+            machine.deletionTime = 0
+            self.models.vmachine.set(machine)
+        gevent.spawn(self.cb.cloudspace.update_firewall, cloudspace)
+        for disk in self.models.disk.search({'id': {'$in': machine.disks}})[1:]:
+            disk['status'] = 'CREATED'
+            self.models.disk.set(disk)
```

Comment:
thi can be done with one `updateSearch({'id' : {'$in': machine.disks}}, {'$set': {'status': 'CREATED'}})`